spec/policy/upstream/upstream_spec.lua

```diff
@@ -64,90 +66,58 @@ describe('Upstream policy', function()
 
       it('does not store a URL in the context', function()
-        upstream:rewrite(context)
-        assert.is_nil(context.new_upstream)
+        policy:rewrite(context)
+        assert.is_nil(context[policy])
       end)
     end)
   end)
 
   describe('.content', function()
-    before_each(function()
-      -- Set headers_sent to false, otherwise, the upstream is not changed
-      ngx.headers_sent = false
-
-      stub(ngx.req, 'set_header')
-      stub(ngx, 'exec')
-    end)
-
     describe('when there is a new upstream in the context', function()
-      before_each(function()
-        context.new_upstream = test_upstream_matched_url
-      end)
-
       it('changes the upstream', function()
-        local upstream = UpstreamPolicy.new({})
+        local upstream = Upstream.new(test_upstream_matched)
+        local policy = UpstreamPolicy.new({})
 
-        upstream:content(context)
+        stub.new(upstream, 'rewrite')
+        stub.new(upstream, 'call')
 
-        assert.equals(test_upstream_matched_proxy_pass, ngx.var.proxy_pass)
-        assert.stub(ngx.req.set_header).was_called_with('Host',
-          test_upstream_matched_host)
-        assert.stub(ngx.exec).was_called_with('@upstream')
-      end)
+        local ctx = { [policy] = upstream }
+        policy:content(ctx)
 
-      it('marks in the context that the upstream has changed', function()
-        local upstream = UpstreamPolicy.new({})
-        upstream:content(context)
-        assert.is_truthy(context.upstream_changed)
+        assert.spy(upstream.rewrite).was_called_with(upstream)
+        assert.spy(upstream.call).was_called_with(upstream, ctx)
       end)
     end)
 
     describe('when there is not a new upstream in the context', function()
-      before_each(function()
-        context.new_upstream = nil
-      end)
-
       it('does not change the upstream', function()
-        local upstream = UpstreamPolicy.new({})
-        upstream:content(context)
-
-        assert.is_nil(ngx.var.proxy_pass)
-        assert.stub(ngx.req.set_header).was_not_called()
-        assert.stub(ngx.exec).was_not_called()
-      end)
-
-      it('does not mark in the context that the upstream has changed', function()
-        local upstream = UpstreamPolicy.new({})
-        upstream:content(context)
-        assert.is_falsy(context.upstream_changed)
+        local policy = UpstreamPolicy.new({})
+        assert.returns_error('no upstream', policy:content(context))
       end)
     end)
   end)
 
   describe('.balancer', function()
     describe('when the upstream has been changed in previous phases', function()
       it('calls the balancer', function()
-        stub(balancer, 'call')
-        local upstream = UpstreamPolicy.new()
-
-        upstream:balancer({ upstream_changed = true })
+        local policy = UpstreamPolicy.new()
 
-        assert.stub(balancer.call).was_called()
+        assert.equal(apicast_balancer.call, policy.balancer)
```

Comment:
This tests that the functions are the same, but not that it was called as the `describe` says.